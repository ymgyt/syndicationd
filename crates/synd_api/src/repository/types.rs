use std::collections::HashMap;

use kvsd::Value;
use serde::{Deserialize, Serialize};
use sqlx::prelude::FromRow;
use synd_feed::types::{Category, FeedUrl, Requirement};

use crate::repository::RepositoryError;

#[derive(Debug, Clone)]
pub struct Feed {
    pub url: String,
    pub title: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct FeedSubscription {
    pub user_id: String,
    pub url: FeedUrl,
    pub requirement: Option<Requirement>,
    pub category: Option<Category<'static>>,
}

#[derive(Serialize, Deserialize, Debug, Default)]
pub struct SubscribedFeeds {
    pub urls: Vec<FeedUrl>,
    pub annotations: Option<HashMap<FeedUrl, FeedAnnotations>>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct FeedAnnotations {
    pub requirement: Option<Requirement>,
    pub category: Option<Category<'static>>,
}

impl TryFrom<Value> for SubscribedFeeds {
    type Error = RepositoryError;

    fn try_from(value: Value) -> Result<Self, Self::Error> {
        serde_json::from_slice(&value).map_err(RepositoryError::internal)
    }
}

impl TryFrom<SubscribedFeeds> for Value {
    type Error = RepositoryError;

    fn try_from(value: SubscribedFeeds) -> Result<Self, Self::Error> {
        let value = serde_json::to_vec(&value).map_err(RepositoryError::internal)?;
        Ok(Value::new(value).unwrap())
    }
}
