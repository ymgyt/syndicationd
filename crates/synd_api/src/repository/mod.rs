mod subscription;
use ::kvsd::KvsdError;
pub use subscription::SubscriptionRepository;

pub mod kvsd;
pub mod sqlite;
pub mod types;

#[derive(thiserror::Error, Debug)]
pub enum RepositoryError {
    #[error("internal error: {0}")]
    Internal(#[from] anyhow::Error),
    #[error(transparent)]
    Migrate(#[from] sqlx::migrate::MigrateError),
}

impl RepositoryError {
    pub fn internal(err: impl Into<anyhow::Error>) -> Self {
        RepositoryError::Internal(err.into())
    }
}

impl From<KvsdError> for RepositoryError {
    fn from(value: KvsdError) -> Self {
        RepositoryError::Internal(value.into())
    }
}

impl From<sqlx::Error> for RepositoryError {
    fn from(value: sqlx::Error) -> Self {
        RepositoryError::internal(anyhow::Error::from(value))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn error() {
        assert!(
            !RepositoryError::internal(anyhow::anyhow!("error"))
                .to_string()
                .is_empty()
        );

        assert!(
            !RepositoryError::from(KvsdError::Unauthenticated)
                .to_string()
                .is_empty()
        );
    }
}
