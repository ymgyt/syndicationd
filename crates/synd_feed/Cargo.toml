[package]
authors.workspace    = true
categories.workspace = true
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
repository.workspace = true

description = "Library to handle syndication spec"
name        = "synd-feed"
readme      = "README.md"
version     = "0.3.5"

[dependencies]
synd-o11y = { path = "../synd_o11y", version = "0.1.9" }

anyhow        = { workspace = true }
async-graphql = { workspace = true, optional = true }
async-trait   = { workspace = true }
chrono        = { workspace = true }
fake          = { workspace = true, optional = true, features = ["derive"] }
feed-rs       = { workspace = true }
futures-util  = { workspace = true }
moka          = { workspace = true, features = ["future"] }
rand          = { workspace = true, optional = true }
reqwest       = { workspace = true, features = ["stream"] }
schemars      = { workspace = true, optional = true, features = ["derive", "url"] }
serde         = { workspace = true }
serde_json    = { workspace = true }
sqlx          = { workspace = true, optional = true, features = ["sqlite"] }
thiserror     = { workspace = true }
tokio         = { workspace = true, features = ["macros"] }
tokio-util    = { workspace = true }
tracing       = { workspace = true }
url           = { workspace = true, features = ["serde"] }

[features]
fake       = ["dep:fake", "dep:rand"]
graphql    = ["dep:async-graphql"]
jsonschema = ["dep:schemars"]
sqlx       = ["dep:sqlx"]

[lints]
workspace = true

[package.metadata.release]
pre-release-hook = ["just", "changelog", "{{version}}"]
pre-release-replacements = [
  { file = "CHANGELOG.md", search = "unreleased", replace = "v{{version}}", min = 0 },
  { file = "CHANGELOG.md", search = "__release_date__", replace = "{{date}}", min = 0 },
  { file = "README.md", search = "/synd-feed-v.*/", replace = "/{{crate_name}}-v{{version}}/", min = 0 },
]
tag-message = "chore: release {{crate_name}} version {{version}}"

[package.metadata.dist]
dist = false
