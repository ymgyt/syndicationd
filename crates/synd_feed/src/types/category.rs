use std::{
    borrow::Cow,
    fmt::{self, Display},
};

use serde::{Deserialize, Serialize};
use thiserror::Error;

#[derive(Error, Debug, PartialEq, Eq)]
pub enum CategoryError {
    #[error("not empty validation is violated")]
    NotEmptyViolated,
    #[error("len max validation is violated")]
    LenMaxViolated,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[cfg_attr(feature = "jsonschema", derive(schemars::JsonSchema))]
pub struct Category<'a>(Cow<'a, str>);

impl<'a> Category<'a> {
    const MAX_LEN: usize = 30;
    pub fn new(c: impl Into<Cow<'a, str>>) -> Result<Self, CategoryError> {
        let c = c.into().trim().to_ascii_lowercase();

        match c.len() {
            0 => return Err(CategoryError::NotEmptyViolated),
            n if n > Self::MAX_LEN => return Err(CategoryError::LenMaxViolated),
            _ => {}
        }

        Ok(Self(c.into()))
    }

    pub fn as_str(&self) -> &str {
        self.0.as_ref()
    }

    pub fn into_inner(self) -> Cow<'a, str> {
        self.0
    }
}

impl<'s> TryFrom<&'s str> for Category<'s> {
    type Error = CategoryError;

    fn try_from(value: &'s str) -> Result<Self, Self::Error> {
        Self::new(value)
    }
}

impl Display for Category<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.pad(self.0.as_ref())
    }
}

#[cfg(feature = "graphql")]
#[async_graphql::Scalar]
impl async_graphql::ScalarType for Category<'_> {
    fn parse(value: async_graphql::Value) -> async_graphql::InputValueResult<Self> {
        let async_graphql::Value::String(s) = value else {
            return Err(async_graphql::InputValueError::expected_type(value));
        };

        match Category::new(s) {
            Ok(c) => Ok(c),
            Err(err) => Err(async_graphql::InputValueError::custom(err)),
        }
    }

    fn to_value(&self) -> async_graphql::Value {
        // Is this clone avoidable?
        async_graphql::Value::String(self.0.clone().into_owned())
    }
}

#[cfg(feature = "fake")]
impl fake::Dummy<fake::Faker> for Category<'_> {
    fn dummy_with_rng<R: rand::Rng + ?Sized>(_config: &fake::Faker, rng: &mut R) -> Self {
        let category: String = fake::Fake::fake_with_rng(&(1..31), rng);
        Self::new(category).unwrap()
    }
}

#[cfg(feature = "sqlx")]
impl sqlx::Type<sqlx::Sqlite> for Category<'static> {
    fn type_info() -> sqlx::sqlite::SqliteTypeInfo {
        <String as sqlx::Type<sqlx::Sqlite>>::type_info()
    }
}

#[cfg(feature = "sqlx")]
impl<'q> sqlx::Encode<'q, sqlx::Sqlite> for Category<'static> {
    fn encode_by_ref(
        &self,
        buf: &mut Vec<sqlx::sqlite::SqliteArgumentValue<'q>>,
    ) -> Result<sqlx::encode::IsNull, Box<dyn std::error::Error + Send + Sync>> {
        self.to_string().encode_by_ref(buf)
    }
}

#[cfg(feature = "sqlx")]
impl<'r> sqlx::Decode<'r, sqlx::Sqlite> for Category<'static> {
    fn decode(value: sqlx::sqlite::SqliteValueRef<'r>) -> Result<Self, sqlx::error::BoxDynError> {
        let s = <String as sqlx::Decode<sqlx::Sqlite>>::decode(value)?;
        Self::new(Cow::Owned(s)).map_err(Into::into)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn category_spec() {
        assert_eq!(Category::new(""), Err(CategoryError::NotEmptyViolated));
        assert_eq!(
            Category::new("a".repeat(Category::MAX_LEN + 1)),
            Err(CategoryError::LenMaxViolated)
        );

        assert!(Category::new("a".repeat(Category::MAX_LEN) + "  ").is_ok(),);

        assert_eq!(
            Category::new("rust").unwrap().into_inner(),
            format!("{}", Category::new("rust").unwrap()),
        );
    }

    #[test]
    #[cfg(feature = "graphql")]
    fn scalar() {
        use async_graphql::ScalarType;

        assert!(Category::parse(async_graphql::Value::Null).is_err());
        assert!(
            Category::parse(async_graphql::Value::String(
                "a".repeat(Category::MAX_LEN + 1)
            ))
            .is_err()
        );
    }

    #[test]
    #[cfg(feature = "fake")]
    fn fake() {
        use fake::Fake;

        let c: Category = fake::Faker.fake();
        assert!(!c.as_str().is_empty());
    }
}
