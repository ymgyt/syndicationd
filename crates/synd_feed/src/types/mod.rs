use std::{borrow::Cow, fmt::Display};

use chrono::{DateTime, Utc};
use feed_rs::model::{self as feedrs, Generator, Link, Person, Text};

pub type Time = DateTime<Utc>;

mod requirement;
pub use requirement::Requirement;

mod category;
pub use category::Category;

mod url;
pub use url::FeedUrl;

mod feed_type;
pub use feed_type::FeedType;

mod macros;

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct EntryId<'a>(Cow<'a, str>);

impl<'a, T> From<T> for EntryId<'a>
where
    T: Into<Cow<'a, str>>,
{
    fn from(value: T) -> Self {
        Self(value.into())
    }
}

impl Display for EntryId<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.0.as_ref())
    }
}

#[derive(Debug, Clone)]
pub struct Entry(feedrs::Entry);

impl Entry {
    pub fn id(&self) -> EntryId<'static> {
        EntryId(Cow::Owned(self.0.id.clone()))
    }

    pub fn id_ref(&self) -> EntryId<'_> {
        EntryId(Cow::Borrowed(self.0.id.as_str()))
    }

    pub fn title(&self) -> Option<&str> {
        self.0.title.as_ref().map(|text| text.content.as_str())
    }

    pub fn updated(&self) -> Option<Time> {
        self.0.updated
    }

    pub fn published(&self) -> Option<Time> {
        self.0.published
    }

    pub fn summary(&self) -> Option<&str> {
        self.0.summary.as_ref().map(|text| text.content.as_str())
    }

    pub fn content(&self) -> Option<&str> {
        self.0
            .content
            .as_ref()
            .and_then(|content| content.body.as_deref())
    }

    pub fn website_url(&self, feed_type: FeedType) -> Option<&str> {
        link::find_website_url(feed_type, &self.0.links)
    }

    /// Return approximate entry bytes size
    pub fn approximate_size(&self) -> usize {
        let content_size = self
            .0
            .content
            .as_ref()
            .and_then(|content| content.body.as_deref())
            .map_or(0, str::len);

        let summary_size = self
            .0
            .summary
            .as_ref()
            .map_or(0, |summary| summary.content.len());

        content_size + summary_size
    }
}

#[derive(Debug, Clone)]
pub struct FeedMeta {
    url: FeedUrl,
    // feed_rs models
    feed_type: FeedType,
    title: Option<Text>,
    updated: Option<Time>,
    authors: Vec<Person>,
    description: Option<Text>,
    links: Vec<Link>,
    generator: Option<Generator>,
    published: Option<Time>,
}

#[derive(Debug, Clone)]
pub struct Annotated<T> {
    pub feed: T,
    pub requirement: Option<Requirement>,
    pub category: Option<Category<'static>>,
}

impl<T> Annotated<T> {
    pub fn project<U>(&self, f: impl Fn(&T) -> U) -> Annotated<U> {
        Annotated {
            feed: f(&self.feed),
            requirement: self.requirement,
            category: self.category.clone(),
        }
    }
}

impl<T> Annotated<T> {
    pub fn new(feed: T) -> Self {
        Self {
            feed,
            requirement: None,
            category: None,
        }
    }
}

impl FeedMeta {
    pub fn r#type(&self) -> FeedType {
        self.feed_type
    }

    pub fn url(&self) -> &FeedUrl {
        &self.url
    }

    pub fn title(&self) -> Option<&str> {
        self.title.as_ref().map(|text| text.content.as_str())
    }

    pub fn updated(&self) -> Option<Time> {
        self.updated.or(self.published)
    }

    pub fn authors(&self) -> impl Iterator<Item = &str> {
        self.authors.iter().map(|person| person.name.as_str())
    }

    pub fn description(&self) -> Option<&str> {
        self.description.as_ref().map(|text| text.content.as_str())
    }

    pub fn links(&self) -> impl Iterator<Item = &feedrs::Link> {
        self.links.iter()
    }

    /// Return website link to which feed syndicate
    pub fn website_url(&self) -> Option<&str> {
        link::find_website_url(self.r#type(), &self.links)
    }

    pub fn generator(&self) -> Option<&str> {
        self.generator.as_ref().map(|g| g.content.as_str())
    }
}

impl<'a> From<&'a FeedMeta> for Cow<'a, FeedMeta> {
    fn from(value: &'a FeedMeta) -> Self {
        Cow::Borrowed(value)
    }
}

impl From<FeedMeta> for Cow<'static, FeedMeta> {
    fn from(value: FeedMeta) -> Self {
        Cow::Owned(value)
    }
}

#[derive(Debug, Clone)]
pub struct Feed {
    meta: FeedMeta,
    entries: Vec<Entry>,
}

impl Feed {
    pub fn parts(self) -> (FeedMeta, Vec<Entry>) {
        (self.meta, self.entries)
    }

    pub fn meta(&self) -> &FeedMeta {
        &self.meta
    }

    pub fn entries(&self) -> impl Iterator<Item = &Entry> {
        self.entries.iter()
    }

    /// Return approximate Feed byte size
    pub fn approximate_size(&self) -> usize {
        self.entries().map(Entry::approximate_size).sum()
    }
}

impl From<(FeedUrl, feed_rs::model::Feed)> for Feed {
    fn from((url, feed): (FeedUrl, feedrs::Feed)) -> Self {
        let feed_rs::model::Feed {
            feed_type,
            title,
            updated,
            authors,
            description,
            links,
            generator,
            published,
            entries,
            ..
        } = feed;
        let meta = FeedMeta {
            url,
            feed_type: feed_type.into(),
            title,
            updated,
            authors,
            description,
            links,
            generator,
            published,
        };
        let entries = entries.into_iter().map(Entry).collect();

        Feed { meta, entries }
    }
}

mod link {
    use feed_rs::model::Link;

    use crate::types::FeedType;

    pub fn find_website_url<'a>(
        feed_type: FeedType,
        links: impl IntoIterator<Item = &'a Link>,
    ) -> Option<&'a str> {
        let mut links = links.into_iter();
        match feed_type {
            // Find rel == alternate link
            FeedType::Atom => links
                .find(|link| link.rel.as_deref() == Some("alternate"))
                .map(|link| link.href.as_str()),

            // how to detect homepage(website) url?
            // ignore .json extension link
            FeedType::JSON => links
                .find(|link| {
                    !std::path::Path::new(link.href.as_str())
                        .extension()
                        .is_some_and(|ext| ext.eq_ignore_ascii_case("json"))
                })
                .map(|link| link.href.as_str()),

            FeedType::RSS0 => {
                tracing::warn!("RSS0 is used! {:?}", links.collect::<Vec<_>>());
                None
            }

            // Use the first link whose rel is not "self"
            FeedType::RSS1 | FeedType::RSS2 => links
                .find(|link| link.rel.as_deref() != Some("self"))
                .map(|link| link.href.as_str()),
        }
    }

    #[cfg(test)]
    mod tests {
        use super::*;

        #[test]
        fn rss_ignore_rel_self() {
            let links = vec![
                Link {
                    href: "https://syndicationd.ymgyt.io/".into(),
                    title: None,
                    rel: None,
                    media_type: None,
                    href_lang: None,
                    length: None,
                },
                Link {
                    href: "https://syndicationd.ymgyt.io/atom.xml".into(),
                    title: None,
                    rel: Some("self".into()),
                    media_type: None,
                    href_lang: None,
                    length: None,
                },
            ];

            assert_eq!(
                find_website_url(FeedType::RSS1, &links),
                Some("https://syndicationd.ymgyt.io/")
            );
            assert_eq!(
                find_website_url(FeedType::RSS2, &links),
                Some("https://syndicationd.ymgyt.io/")
            );
        }

        #[test]
        fn atom_use_rel_alternate() {
            let links = vec![
                Link {
                    href: "https://syndicationd.ymgyt.io/atom.xml".into(),
                    title: None,
                    rel: Some("self".into()),
                    media_type: None,
                    href_lang: None,
                    length: None,
                },
                Link {
                    href: "https://syndicationd.ymgyt.io/".into(),
                    title: None,
                    rel: Some("alternate".into()),
                    media_type: None,
                    href_lang: None,
                    length: None,
                },
            ];

            assert_eq!(
                find_website_url(FeedType::Atom, &links),
                Some("https://syndicationd.ymgyt.io/")
            );
        }

        #[test]
        fn json_ignore_json_ext() {
            let links = vec![
                Link {
                    href: "https://kubernetes.io/docs/reference/issues-security/official-cve-feed/index.json".into(),
                    title: None,
                    rel: None,
                    media_type: None,
                    href_lang: None,
                    length: None,
                },
                Link {
                    href: "https://kubernetes.io".into(),
                    title: None,
                    rel: None,
                    media_type: None,
                    href_lang: None,
                    length: None,
                },
            ];

            assert_eq!(
                find_website_url(FeedType::JSON, &links),
                Some("https://kubernetes.io")
            );
        }
    }
}
