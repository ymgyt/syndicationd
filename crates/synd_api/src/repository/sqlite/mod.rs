use std::{collections::HashMap, path::Path};

use async_trait::async_trait;
use sqlx::{SqlitePool, sqlite::SqliteConnectOptions};
use tracing::info;

use crate::repository::{
    self, RepositoryError, SubscriptionRepository,
    subscription::RepositoryResult,
    types::{FeedAnnotations, SubscribedFeeds},
};

pub struct DbPool {
    pool: SqlitePool,
}

impl DbPool {
    pub async fn connect(db_path: impl AsRef<Path>) -> Result<Self, RepositoryError> {
        let opts = SqliteConnectOptions::new()
            .filename(db_path)
            .create_if_missing(true)
            .foreign_keys(true);

        Self::do_connect(opts).await
    }

    pub async fn migrate(&self) -> Result<(), RepositoryError> {
        info!("Run migrations...");
        sqlx::migrate!().run(&self.pool).await?;
        Ok(())
    }

    async fn do_connect(opts: SqliteConnectOptions) -> Result<Self, RepositoryError> {
        info!(?opts, "Connecting to sqlite...");
        let pool = SqlitePool::connect_with(opts).await?;

        // TODO: configure pool options

        Ok(DbPool { pool })
    }

    #[cfg(test)]
    pub async fn in_memory() -> Result<Self, RepositoryError> {
        use sqlx::sqlite::SqlitePoolOptions;

        let opts = SqliteConnectOptions::new()
            .in_memory(true)
            .foreign_keys(true);
        let pool = SqlitePoolOptions::new()
            .max_connections(1)
            .connect_with(opts)
            .await?;
        Ok(DbPool { pool })
    }
}

#[async_trait]
impl SubscriptionRepository for DbPool {
    #[tracing::instrument(name = "repo::put_feed_subscription", skip_all)]
    async fn put_feed_subscription(
        &self,
        feed: repository::types::FeedSubscription,
    ) -> RepositoryResult<()> {
        let feed_url = feed.url.to_string();
        let requirement = feed.requirement.map(|r| r.to_string());
        let category = feed.category.map(|c| c.to_string());

        sqlx::query!(
            r#"
            INSERT INTO subscribed_feed (user_id, url, requirement, category)
            VALUES (?, ?, ?, ?)
            ON CONFLICT(user_id, url) DO UPDATE SET
                requirement = excluded.requirement,
                category = excluded.category
            "#,
            feed.user_id,
            feed_url,
            requirement,
            category,
        )
        .execute(&self.pool)
        .await
        .map_err(RepositoryError::internal)?;

        Ok(())
    }

    async fn delete_feed_subscription(
        &self,
        feed: repository::types::FeedSubscription,
    ) -> RepositoryResult<()> {
        let affected = sqlx::query!(
            r#"
            DELETE FROM subscribed_feed
            WHERE user_id = ? AND url = ?
            "#,
            feed.user_id,
            feed.url,
        )
        .execute(&self.pool)
        .await
        .map_err(RepositoryError::internal)?
        .rows_affected();

        if affected > 0 {
            info!(url = %feed.url, "Delete subscribed feed");
        }

        Ok(())
    }

    async fn fetch_subscribed_feeds(&self, user_id: &str) -> RepositoryResult<SubscribedFeeds> {
        use synd_feed::types::{Category, FeedUrl, Requirement};
        let feeds = sqlx::query_as!(
            repository::types::FeedSubscription,
            r#"
                SELECT
                    user_id,
                    url AS "url: FeedUrl",
                    requirement AS "requirement: Requirement",
                    category AS "category: Category"
                FROM subscribed_feed
                WHERE user_id = ?
                ORDER BY created_at DESC
            "#,
            user_id
        )
        .fetch_all(&self.pool)
        .await?
        .into_iter()
        .fold(
            SubscribedFeeds {
                urls: Vec::new(),
                annotations: Some(HashMap::new()),
            },
            |mut feeds, feed| {
                let annot = FeedAnnotations {
                    requirement: feed.requirement,
                    category: feed.category,
                };
                feeds
                    .annotations
                    .as_mut()
                    .unwrap()
                    .insert(feed.url.clone(), annot);
                feeds.urls.push(feed.url);
                feeds
            },
        );
        tracing::info!("{feeds:?}");
        Ok(feeds)
    }
}

#[cfg(test)]
mod tests {
    use synd_feed::types::{Category, Requirement};

    use super::*;

    async fn test_db() -> DbPool {
        let db = DbPool::in_memory().await.unwrap();
        db.migrate().await.unwrap();
        db
    }

    #[tokio::test]
    async fn feed_subscription() -> anyhow::Result<()> {
        let db = test_db().await;
        let user_id = String::from("test1");

        let mut test_feeds = vec![
            repository::types::FeedSubscription {
                user_id: user_id.clone(),
                url: "https://ymgyt.io/feed_1".try_into().unwrap(),
                requirement: Some(Requirement::Must),
                category: Some(Category::new("rust").unwrap()),
            },
            repository::types::FeedSubscription {
                user_id: user_id.clone(),
                url: "https://ymgyt.io/feed_2".try_into().unwrap(),
                requirement: Some(Requirement::Should),
                category: Some(Category::new("linux").unwrap()),
            },
        ];

        let feeds = db.fetch_subscribed_feeds(&user_id).await?;
        assert!(feeds.urls.is_empty());

        // create
        {
            for feed in &test_feeds {
                db.put_feed_subscription(feed.clone()).await?;
            }
            let feeds = db.fetch_subscribed_feeds(&user_id).await?;

            insta::assert_yaml_snapshot!("create", feeds, {
                ".annotations" => insta::sorted_redaction(),
            });
        }

        // update
        {
            test_feeds[0].requirement = Some(Requirement::May);
            test_feeds[0].category = Some(Category::new("foo").unwrap());
            db.put_feed_subscription(test_feeds[0].clone()).await?;

            let feeds = db.fetch_subscribed_feeds(&user_id).await?;
            insta::assert_yaml_snapshot!("update", feeds, {
                ".annotations" => insta::sorted_redaction(),
            });
        }

        // delete
        {
            db.delete_feed_subscription(test_feeds[0].clone()).await?;
            let feeds = db.fetch_subscribed_feeds(&user_id).await?;
            insta::assert_yaml_snapshot!("delete", feeds, {
                ".annotations" => insta::sorted_redaction(),
            });
        }

        Ok(())
    }
}
