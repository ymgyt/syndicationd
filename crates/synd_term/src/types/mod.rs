use chrono::DateTime;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use synd_feed::types::{Category, FeedType, FeedUrl, Requirement};

use crate::{
    client::{
        mutation,
        query::{self},
    },
    ui,
};

mod time;
pub use time::{Time, TimeExt};

mod page_info;
pub use page_info::PageInfo;

mod requirement_ext;
pub use requirement_ext::RequirementExt;

pub(crate) mod github;

#[derive(Debug, Clone)]
#[cfg_attr(test, derive(fake::Dummy))]
pub struct Link {
    pub href: String,
    pub rel: Option<String>,
    pub media_type: Option<String>,
    pub title: Option<String>,
}

impl From<query::subscription::Link> for Link {
    fn from(v: query::subscription::Link) -> Self {
        Self {
            href: v.href,
            rel: v.rel,
            media_type: v.media_type,
            title: v.title,
        }
    }
}

impl From<mutation::subscribe_feed::Link> for Link {
    fn from(v: mutation::subscribe_feed::Link) -> Self {
        Self {
            href: v.href,
            rel: v.rel,
            media_type: v.media_type,
            title: v.title,
        }
    }
}

#[derive(Debug, Clone)]
#[cfg_attr(test, derive(fake::Dummy))]
pub struct EntryMeta {
    pub title: Option<String>,
    pub published: Option<Time>,
    pub updated: Option<Time>,
    pub summary: Option<String>,
}

impl From<query::subscription::EntryMeta> for EntryMeta {
    fn from(e: query::subscription::EntryMeta) -> Self {
        Self {
            title: e.title,
            published: e.published.map(parse_time),
            updated: e.updated.map(parse_time),
            summary: e.summary,
        }
    }
}

impl From<mutation::subscribe_feed::EntryMeta> for EntryMeta {
    fn from(e: mutation::subscribe_feed::EntryMeta) -> Self {
        Self {
            title: e.title,
            published: e.published.map(parse_time),
            updated: e.updated.map(parse_time),
            summary: e.summary,
        }
    }
}

impl EntryMeta {
    pub fn summary_text(&self, width: usize) -> Option<String> {
        self.summary
            .as_deref()
            .map(|summary| html2text::from_read(summary.as_bytes(), width))
    }
}

#[derive(Debug, Clone)]
#[cfg_attr(test, derive(fake::Dummy))]
pub struct Feed {
    pub feed_type: Option<FeedType>,
    pub title: Option<String>,
    pub url: FeedUrl,
    pub updated: Option<Time>,
    pub links: Vec<Link>,
    pub website_url: Option<String>,
    pub description: Option<String>,
    pub generator: Option<String>,
    pub entries: Vec<EntryMeta>,
    pub authors: Vec<String>,
    requirement: Option<Requirement>,
    category: Option<Category<'static>>,
}

impl Feed {
    pub fn requirement(&self) -> Requirement {
        self.requirement.unwrap_or(ui::DEFAULT_REQUIREMNET)
    }

    pub fn category(&self) -> &Category<'static> {
        self.category.as_ref().unwrap_or(ui::default_category())
    }

    #[must_use]
    pub fn with_url(self, url: FeedUrl) -> Self {
        Self { url, ..self }
    }

    #[must_use]
    pub fn with_requirement(self, requirement: Requirement) -> Self {
        Self {
            requirement: Some(requirement),
            ..self
        }
    }

    #[must_use]
    pub fn with_category(self, category: Category<'static>) -> Self {
        Self {
            category: Some(category),
            ..self
        }
    }
}

impl From<query::subscription::Feed> for Feed {
    fn from(f: query::subscription::Feed) -> Self {
        Self {
            feed_type: match f.type_ {
                query::subscription::FeedType::ATOM => Some(FeedType::Atom),
                query::subscription::FeedType::RSS1 => Some(FeedType::RSS1),
                query::subscription::FeedType::RSS2 => Some(FeedType::RSS2),
                query::subscription::FeedType::RSS0 => Some(FeedType::RSS0),
                query::subscription::FeedType::JSON => Some(FeedType::JSON),
                query::subscription::FeedType::Other(_) => None,
            },
            title: f.title,
            url: f.url,
            updated: f.updated.map(parse_time),
            links: f.links.nodes.into_iter().map(From::from).collect(),
            website_url: f.website_url,
            description: f.description,
            generator: f.generator,
            entries: f.entries.nodes.into_iter().map(From::from).collect(),
            authors: f.authors.nodes,
            requirement: f.requirement.and_then(|r| match r {
                query::subscription::Requirement::MUST => Some(Requirement::Must),
                query::subscription::Requirement::SHOULD => Some(Requirement::Should),
                query::subscription::Requirement::MAY => Some(Requirement::May),
                query::subscription::Requirement::Other(_) => None,
            }),
            category: f.category,
        }
    }
}

impl From<mutation::subscribe_feed::Feed> for Feed {
    fn from(f: mutation::subscribe_feed::Feed) -> Self {
        Self {
            feed_type: match f.type_ {
                mutation::subscribe_feed::FeedType::ATOM => Some(FeedType::Atom),
                mutation::subscribe_feed::FeedType::RSS1 => Some(FeedType::RSS1),
                mutation::subscribe_feed::FeedType::RSS2 => Some(FeedType::RSS2),
                mutation::subscribe_feed::FeedType::RSS0 => Some(FeedType::RSS0),
                mutation::subscribe_feed::FeedType::JSON => Some(FeedType::JSON),
                mutation::subscribe_feed::FeedType::Other(_) => None,
            },
            title: f.title,
            url: f.url,
            updated: f.updated.map(parse_time),
            links: f.links.nodes.into_iter().map(From::from).collect(),
            website_url: f.website_url,
            description: f.description,
            generator: f.generator,
            entries: f.entries.nodes.into_iter().map(From::from).collect(),
            authors: f.authors.nodes,
            requirement: f.requirement.and_then(|r| match r {
                mutation::subscribe_feed::Requirement::MUST => Some(Requirement::Must),
                mutation::subscribe_feed::Requirement::SHOULD => Some(Requirement::Should),
                mutation::subscribe_feed::Requirement::MAY => Some(Requirement::May),
                mutation::subscribe_feed::Requirement::Other(_) => None,
            }),
            category: f.category,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Entry {
    pub title: Option<String>,
    pub published: Option<Time>,
    pub updated: Option<Time>,
    pub website_url: Option<String>,
    pub summary: Option<String>,
    pub feed_title: Option<String>,
    pub feed_url: FeedUrl,
    requirement: Option<Requirement>,
    category: Option<Category<'static>>,
}

impl Entry {
    pub fn summary_text(&self, width: usize) -> Option<String> {
        self.summary.as_deref().map(|summary| {
            html2text::config::plain()
                .string_from_read(summary.as_bytes(), width)
                .unwrap_or_default()
        })
    }

    pub fn requirement(&self) -> Requirement {
        self.requirement.unwrap_or(ui::DEFAULT_REQUIREMNET)
    }

    pub fn category(&self) -> &Category<'static> {
        self.category
            .as_ref()
            .unwrap_or_else(|| ui::default_category())
    }
}

impl From<query::entries::Entry> for Entry {
    fn from(v: query::entries::Entry) -> Self {
        Self {
            title: v.title,
            published: v.published.map(parse_time),
            updated: v.updated.map(parse_time),
            website_url: v.website_url,
            feed_title: v.feed.title,
            feed_url: v.feed.url,
            summary: v.summary,
            requirement: match v.feed.requirement {
                Some(query::entries::Requirement::MUST) => Some(Requirement::Must),
                Some(query::entries::Requirement::SHOULD) => Some(Requirement::Should),
                Some(query::entries::Requirement::MAY) => Some(Requirement::May),
                _ => None,
            },
            category: v.feed.category,
        }
    }
}

#[derive(Serialize, Deserialize, JsonSchema)]
pub struct ExportedFeed {
    pub title: Option<String>,
    pub url: FeedUrl,
    pub requirement: Option<Requirement>,
    pub category: Option<Category<'static>>,
}

impl From<query::export_subscription::ExportSubscriptionOutputFeedsNodes> for ExportedFeed {
    fn from(v: query::export_subscription::ExportSubscriptionOutputFeedsNodes) -> Self {
        Self {
            title: v.title,
            url: v.url,
            requirement: v.requirement.and_then(|r| match r {
                query::export_subscription::Requirement::MUST => Some(Requirement::Must),
                query::export_subscription::Requirement::SHOULD => Some(Requirement::Should),
                query::export_subscription::Requirement::MAY => Some(Requirement::May),
                query::export_subscription::Requirement::Other(_) => None,
            }),
            category: v.category,
        }
    }
}

impl From<ExportedFeed> for mutation::subscribe_feed::SubscribeFeedInput {
    fn from(feed: ExportedFeed) -> Self {
        Self {
            url: feed.url,
            requirement: feed.requirement.map(|r| match r {
                Requirement::Must => mutation::subscribe_feed::Requirement::MUST,
                Requirement::Should => mutation::subscribe_feed::Requirement::SHOULD,
                Requirement::May => mutation::subscribe_feed::Requirement::MAY,
            }),
            category: feed.category,
        }
    }
}

fn parse_time(t: impl AsRef<str>) -> Time {
    DateTime::parse_from_rfc3339(t.as_ref())
        .expect("invalid rfc3339 time")
        .with_timezone(&chrono::Utc)
}
