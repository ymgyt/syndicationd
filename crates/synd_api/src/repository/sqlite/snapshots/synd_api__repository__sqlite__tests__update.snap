---
source: crates/synd_api/src/repository/sqlite/mod.rs
expression: feeds
---
urls:
  - "https://ymgyt.io/feed_1"
  - "https://ymgyt.io/feed_2"
annotations:
  "https://ymgyt.io/feed_1":
    requirement: May
    category: foo
  "https://ymgyt.io/feed_2":
    requirement: Should
    category: linux
